index.js

```diff
@@ -7,27 +7,73 @@
 
 'use strict';
 
-module.exports = function(prefix) {
-  var debug = require('debug');
-
-  return function(app) {
-    this.define('debug', function(name) {
-      var fn = debug(namespace(prefix, name));
-      return function() {
-        return fn.apply(debug, arguments);
-      };
+/**
+ * Plugin that adds `debug` method to the instance
+ * and `namespaces` are added to this method.
+ *
+ * ```js
+ * app.debug()('debugging on main namespace')
+ * //=> base debugging on main namespace
+ *
+ * app.debug.one = app.debug('one')
+ * app.debug.one('testing %s, whatever', 'foo, bar')
+ * //=> base:one testing foo, bar, whatever
+ *
+ * app.debug.mix = app.debug('two', 123 'three:four', 'five')
+ * app.debug.mix('okey, %s awesome', 'this is')
+ * //=> base:two:three:four:five okey, this is awesome
+ *
+ * app.debug.helper('is one of internal namespaces added')
+ * //=> base:helper is one of internal namespaces added
+ * ```
+ *
+ * @name   debug
+ * @param  {Array} `namespaces`
+ * @return {Function}
+ */
+
+module.exports = function debugPlugin(namespaces) {
+  namespaces = Array.isArray(namespaces) ? namespaces : [namespaces];
+
+  return function plugin(app) {
+    app.define('_debugPrefix', app._namespace);
+    app.define('debug', function debug() {
+      return debugFactory.apply(app, arguments);
     });
+    if (namespaces.length) {
+      var len = namespaces.length;
+      var i = 0;
+
+      while (i < len) {
+        var ns = namespaces[i++];
+        app.debug[ns] = debugFactory.call(app, ns);
+        app.debug[ns].color = app.debug.color;
+      }
+    }
```

Comment:
Yea.. it's.. damn, you're not a human being, :laughing: 
I was a bit familiar with it, but didn't realize it is so big and awesome, and that it can be so nested.
Great clarification.